src/dmd/utils.d

```diff
@@ -138,3 +139,44 @@ const(char)[] toDString (const(char)* s) pure nothrow @nogc
 {
     return s ? s[0 .. strlen(s)] : null;
 }
+
+/**
+   Copy the content of `src` into a C-string ('\0' terminated) then call `dg`
+
+   The intent of this function is to provide an allocation-less
+   way to call a C function using a D slice.
+   The function internally allocates a buffer if needed, but frees it on exit.
+
+   Note:
+     The argument to `dg` is `scope`. To keep the data around after `dg` exits,
+     one has to copy it.
+
+   Params:
+     src = Slice to use to call the C function
+     dg  = Delegate to call afterwards
+
+   Returns:
+     The return value of `T`
+*/
+auto toCStringThen(alias dg)(const(char)[] src) nothrow
+{
+    char[512] small = void;
+    scope ptr = (src.length < (small.length - 1))
+                    ? small[0 .. src.length + 1]
+                    : (cast(char*)mem.xmalloc(src.length + 1))[0 .. src.length + 1];
+    scope (exit)
+    {
+        if (&ptr[0] != &small[0])
+            mem.xfree(&ptr[0]);
+    }
+    ptr[0 .. src.length] = src[];
+    ptr[src.length] = '\0';
+    return dg(ptr);
+}
+
+unittest
+{
+    assert("Hello world".toCStringThen!((v) => v == "Hello world\0"));
+    assert("Hello world\0".toCStringThen!((v) => v == "Hello world\0\0"));
```

Comment:
Why adding an additional null character if one is already present?

Reply:
This function does not care about the content of the slice. And slices should not contain the terminating null char, it should be out of bounds.

Reply:
But all static D strings already do contain a null afterwards, e.g. https://run.dlang.io/is/Abty1m so it looks like we could save quite a bit of copying here.

Reply:
>  e.g. https://run.dlang.io/is/Abty1m

Your example is reading past the end of the slice. It's unsafe and cannot be relied on in this function.
Also that copy is unlikely to have any noticeable performance impact, since this function will be called  primarily to wrap C function calls, and we'll be calling mostly I/O functions.